File: starter-package/creating-test-doubles-with-mockito/creating-spies-with-mockito/src/main/java/com/testwithspring/task/TaskRepository.java
package com.testwithspring.task;

import java.util.Optional;

interface TaskRepository {

    /**
     * @return The number of tasks found from the database.
     */
    long count();

    /**
     * Deletes task from the database.
     * @param task  The deleted task.
     */
    void delete(Task task);

    /**
     * Finds a task by using its id as a search criteria.
     * @param id    The id of the requested task.
     * @return      An {@code Optional} object that contains the found task.
     *              If no task is found, this method returns an empty
     *              {@code Optional} object.
     */
    Optional<Task> findById(Long id);

    /**
     * Saves a task in the database.
     * @param task  The saved task.
     * @return      The saved task.
     */
    Task save(Task task);
}

File: master-package/bonus-webinars-2019/unit-tests-rest-api/src/main/kotlin/com/testwithspring/master/task/TaskRepository.kt
package com.testwithspring.master.task

import org.springframework.stereotype.Repository
import org.springframework.transaction.annotation.Transactional
import java.lang.RuntimeException

/**
 * This repository isn't meant to be used in product.
 * This repository exists only because it allows us to
 * compile our example application and write unit tests
 * for it.
 */
@Repository
open class TaskRepository {

    /**
     * Inserts a new task into the database.
     * @param   input   The information of the created task.
     * @return  The information of the created task.
     */
    @Transactional
    open fun create(input: CreateTask): Task {
        throw RuntimeException("not implemented")
    }

    /**
     * Finds the information of all tasks found from the database.
     * @return  A list of found tasks. If no tasks are found, this
     *          function returns an empty list.
     */
    @Transactional(readOnly = true)
    open fun findAll(): List<TaskListItemDTO> {
        return listOf()
    }

    /**
     * Finds the information of the specified task.
     * @param   id  The id of the requested task.
     * @return  This function returns {@code null}
     */
    @Transactional(readOnly = true)
    open fun findById(id: Long): Task? {
        return null
    }
}